text/0001-single-page-applications.md

```diff
@@ -0,0 +1,34 @@
+# Single Page Applications
+- Start Date: 04/23/2019
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/1
+
+## Decision, including impact on distributions
+Core OpenMRS frontend modules will be implemented within a master single page application. Distributions are encouraged
+to add modules to the master single page application. Older OpenMRS modules implemented with JSPs and GSPs are not addressed in this
+proposal, but will be covered in another one.
+
+## Definition
+A [single page application (SPA)](https://en.wikipedia.org/wiki/Single-page_application) is a frontend application that does not
```

Comment:
We also had what we call user interface extension points where modules would extend the default user interface at various points as needed by means of server side java files. An example would be extending the patient dashboard with more tabs. Do you think this will blend very well with SPA? Am almost sure it will. So just a sanity check. 😊  https://wiki.openmrs.org/x/sA8z

Reply:
Hmm, not sure about this one. One of the biggest differences between server rendered JSPs/GSPs and client rendered SPAs is that the modularity/extensibility is happening *in the browser* instead of in the Java code.

Reading through the Module Extension Points wiki that you linked to, those are very much tied to JSPs and are run on the server, not in the browser. Each JSP is its own HTML file, but in an SPA architecture there is only one HTML file and it's not doing any complex rendering on the server side because it lets the browser-side javascript do all the work.

So I think that user interface extension points (as explained in the wiki) would not be part of the new architecture, but rather something that would only continue to be supported for older code. This is related to the **big** question of what to do with the JSPs and GSPs in terms of a migration path. I intentionally left that out of this RFC PR in favor of trying to tackle that one in a subsequent PR.

Reply:
Do you plan having an equivalent of extension points from the browser side?

Reply:
Let me make sure I am understanding Module Extension Points -- they allow you to modify the HTML that gets sent down for a specific page?

There will not be an exact equivalent to these. The proposed architecture is one where there are core OpenMRS microfrontends that distributions may choose to use, and then additional microfrontends that are developed by individual distributions. The ability to pick and choose which microfrontends you want, and to add in your own microfrontends, is what I have been thinking as the main way to customize/extend the OpenMRS frontend. It sounds like Module Extension Points is a way to **modify** core OpenMRS frontend JSP pages, not just pick and choose which ones to use. In the new architecture, I do not have a proposal in mind for letting people modify a microfrontend's code on-the-fly -- I was only thinking in terms of allowing them to pick and choose which microfrontends to include or not, and then add their own microfrontends.

Does that make sense?

Reply:
Yes it makes sense.
Is this just hard to accomplish with the proposed architecture?

Reply:
An example of how we currently use extension points:

On my distro's clinician dashboard (or any dashboard) say I want to display a set of patient summary widgets in some sort of order.. to do this currently one can use some json config to "inject" those widgets into a specific "patient dashboard" extension point.  So in the new model, suppose you have an "current lab test" widget, a "current medications" widget, and a "current vitals" widget, and they are provided by different microfrontends, would there be a mechanism for including them on a single "dashboard" page?

A simpler extension point example is "Visit Actions" and "Actions" extension which is basically a list of actions that a user can take in the current context, and generally each action maps to a link... continuing from the example above you might have have "Order Tests", "Order Labs" and "Enter Vitals" actions that take the user to specific microfrontends for ordering test, labs, entering vitals, etc... ie we are not trying to combine the display from multiple front ends on a single page, but rather allowing the user to navigate to UIs provided by different microfrontends within a certain patient context.

Reply:
> Is this just hard to accomplish with the proposed architecture?

Module extension points seem to be "throw in some javascript and html into a jsp," which means that you can do almost anything with them but also that there's no structure to how they interact and it is harder to ensure that they work together when there are multiple extensions within a single jsp.

With the new architecture, you still can "throw in some javascript and html into the master single page application," but we will be encouraging people to follow the structure and architecture in this RFC instead of reaching into other modules dom and doing stuff with it. There will be ways to configure and modify the behavior of core pages and code.

@mogoodrich thanks for the examples of extension points.

The answer to how to accomplish that kind of thing in the new architecture is different depending on whether it is modifying an existing page that is in openmrs core versus adding a new page.

**Modifying openmrs-core functionality**: For this, I think there's some exploration to do. I'm open to ideas/proposals. One option would be to pass in [single spa custom props](https://single-spa.js.org/docs/building-applications.html#custom-props) to the core modules that change their behavior, or even add your components to the module.

**Adding brand new functionality**: For this, distributions will be able to author their own [single-spa applications](https://single-spa.js.org/docs/building-applications.html) that can take over specific routes. Then the distribution's application code can fully control that page, perhaps reusing some prebuilt components exposed by OpenMRS.

Does that better explain what this proposed architecture will offer?